include/execution.hpp

```diff
@@ -117,56 +119,125 @@ namespace std::execution {
   // [execution.senders.traits]
   namespace __sender_base {
     struct sender_base {};
   }
   using __sender_base::sender_base;
 
-  template <bool _SendsDone, class... _Ts>
-    struct __sender_of {
-      template <template <class...> class _Tuple, template <class...> class _Variant>
-        using value_types = _Variant<_Tuple<_Ts...>>;
-      template <template <class...> class _Variant>
-        using error_types = _Variant<exception_ptr>;
-      static constexpr bool sends_done = _SendsDone;
-    };
+  inline namespace __sender_traits {
+    namespace __impl {
+      struct __default_context {
+        friend void tag_invoke(set_value_t, __default_context&&, auto&&...) {}
+        friend void tag_invoke(set_error_t, __default_context&&, auto&&) noexcept {}
+        friend void tag_invoke(set_done_t, __default_context&&) noexcept {}
+      };
 
-  template <class _Sender>
-    struct __typed_sender {
-      template <template <class...> class _Tuple, template <class...> class _Variant>
-        using value_types = typename _Sender::template value_types<_Tuple, _Variant>;
-      template <template <class...> class _Variant>
-        using error_types = typename _Sender::template error_types<_Variant>;
-      static constexpr bool sends_done = _Sender::sends_done;
-    };
+      template <class _Sender>
+        struct __typed_sender {
+          template <template <class...> class _Tuple, template <class...> class _Variant>
+            using value_types = typename _Sender::template value_types<_Tuple, _Variant>;
+          template <template <class...> class _Variant>
+            using error_types = typename _Sender::template error_types<_Variant>;
+          static constexpr bool sends_done = _Sender::sends_done;
+        };
 
-  template <class _Sender>
-  auto __sender_traits_base_fn() {
-    if constexpr (__has_sender_types<_Sender>) {
-      return __typed_sender<_Sender>{};
-    } else if constexpr (derived_from<_Sender, sender_base>) {
-      return sender_base{};
-    } else if constexpr (__awaitable<_Sender>) {
-      if constexpr (is_void_v<__await_result_t<_Sender>>) {
-        return __sender_of<false>{};
-      } else {
-        return __sender_of<false, __await_result_t<_Sender>>{};
-      }
-    } else {
-      struct __no_sender_traits{
-        using __this_is_not_a_sender = _Sender;
+      struct get_sender_traits_t {
+        template <class _Sender, class _Receiver>
+        constexpr auto operator()(_Sender&&, _Receiver&&) const noexcept {
+          static_assert(sizeof(_Sender), "Incomplete type used with get_sender_traits");
+          static_assert(sizeof(_Receiver), "Incomplete type used with get_sender_traits");
+          if constexpr (tag_invocable<get_sender_traits_t, _Sender, _Receiver>) {
+            if constexpr (is_void_v<tag_invoke_result_t<get_sender_traits_t, _Sender, _Receiver>>) {
+              return sender_base{};
+            } else {
```

Comment:
I guess it comes down to whether we want non-typed senders to be a thing.

Reply:
My preference would be to not support non-typed senders and just make `get_sender_traits_t` part of the base sender concept.

A lot of algorithms would suddenly need to start having heap-allocating versions if they were to support non-typed senders. This would then complicate a lot of the specification of them as presumably we'd also want some way to customise the allocation of that extra state and then also specify what happens if the allocation fails.
You'd basically need to write each algorithm twice (or possibly more combinations, e.g. if takes as input more than one sender).

For some algorithms like `when_all()` I don't think it's even possible to implement it if we don't know what types the input senders are going to produce in advance - we can't form a cross-product of the possible argument types since there is no single place that knows about all of the possibilities.